opt_adapt/opt.py

```diff
@@ -132,13 +148,162 @@ def _gradient_descent(it, forward_run, m, params, u, u_, dJ_, Rspace=False):
+
+    if a_ is None or b_ is None:
+        a_ = fd.Function(dJ.function_space())
+        b_ = fd.Function(dJ.function_space())
+
+    # Find the descent direction
+    if Rspace:
+        dJ_square = fd.Function(dJ).assign(pow(float(dJ), 2))
+        a = fd.Function(dJ).assign(beta_1 * a_ + (1 - beta_1) * dJ)
+        b = fd.Function(dJ).assign(beta_2 * b_ + (1 - beta_2) * dJ_square)
+        a_hat = fd.Function(dJ).assign(a / (1 - pow(beta_1, it)))
+        b_hat = fd.Function(dJ).assign(b / (1 - pow(beta_2, it)))
+        P = fd.Function(dJ).assign(-1 * a_hat / (pow(b_hat, 0.5) + epsilon))
+    else:
+        dJ_square = params.transfer_fn(dJ * dJ, dJ.function_space())
+        a = params.transfer_fn(beta_1 * a_ + (1 - beta_1) * dJ, dJ.function_space())
+        b = params.transfer_fn(
+            beta_2 * b_ + (1 - beta_2) * dJ_square, dJ.function_space()
+        )
+        a_hat = params.transfer_fn(a / (1 - pow(beta_1, it)), dJ.function_space())
+        b_hat = params.transfer_fn(b / (1 - pow(beta_2, it)), dJ.function_space())
+        P = params.transfer_fn(
+            -1 * a_hat / (pow(b_hat, 0.5) + epsilon), dJ.function_space()
+        )
+
+    # Find step length and take a step downhill
+    lr = line_search(forward_run, m, u, P, J, dJ, params, Rspace)
+    u += lr * P
+
+    yield {"lr": lr, "u+": u, "a-": a, "b-": b}
+
+
+def _LBFGS(it, forward_run, m, params, u, rho, s, y, n=5, Rspace=False):
+    """
+    Take one iteration of L-BFGS algorithm.
+
+    :arg it: the current iteration number
+    :arg forward_run: a Python function that
+        implements the forward model and
+        computes the objective functional
+    :arg m: the current mesh
+    :kwarg params: :class:`OptAdaptParameters` instance
+        containing parameters for the optimisation and
+        adaptation routines
+    :arg u: the current control value
+    :arg rho: the list to store previous n rhok values
+    :arg s: the list to store previous n sk values
+    :arg y: the list to store previous n yk values
+    :arg n: the history size of rho, s, y
+    :kwarg Rspace: is the prognostic function
+        space of type 'Real'?
+    """
+
+    # Annotate the tape and compute the gradient
+    J_, u_ = forward_run(m, u, **params.model_options)
+    dJ_ = fd_adj.compute_gradient(J_, fd_adj.Control(u_))
+    yield {"J": J_, "u": u_.copy(deepcopy=True), "dJ": dJ_.copy(deepcopy=True)}
+
+    if s is None or y is None or rho is None:
+        rho = []
+        s = []
+        y = []
+
+    # Update the descent direction P
+    n_ = len(s)
+    q = dJ_.copy(deepcopy=True)
+    if n_ == 0:
+        H = Matrix(dJ_.function_space())
+        P = H.multiply(q)
+    else:
+        a = np.empty((n_,))
+        for i in range(n_ - 1, -1, -1):
+            a[i] = rho[i] * np.dot(s[i].dat.data, q.dat.data)
+            q -= a[i] * y[i]
+        H = Matrix(dJ_.function_space()).scale(
+            np.dot(s[-1].dat.data, y[-1].dat.data)
```

Comment:
It might be useful to define a function
```
def dotproduct(f, g):
    ...
```
to avoid you having to write this over and over again.